### PR TITLE
[Paddle-TRT] Add check for TRT runtime dynamic shape

### PR DESCRIPTION
**Pull request by cryoco**
### PR types

Function optimization
### PR changes

Others
### Describe

explicitly hints how runtime input shapes of TRT engines are inconsistent with min/max input shapes configured in  `SetTRTDynamicShapeInfo()`. The error message is like:
image

### CODE REVIEW

File: paddle/fluid/operators/tensorrt/tensorrt_engine_op.h [hunk 1]
```diff
@@ -76,6 +76,53 @@ static void RuntimeStaticShapeCheck(std::vector<int64_t> runtime_input_shape,
           model_input_shape_str, runtime_input_shape_str));
 }
 
+static void RuntimeDynamicShapeCheck(std::string x,
```
Comment (shangzhizhou) [inline paddle/fluid/operators/tensorrt/tensorrt_engine_op.h]: 输入参数用const &

File: paddle/fluid/operators/tensorrt/tensorrt_engine_op.h [hunk 2]
```diff
@@ -76,6 +76,53 @@ static void RuntimeStaticShapeCheck(std::vector<int64_t> runtime_input_shape,
...
+                                     std::vector<int64_t> runtime_input_shape,
+                                     std::vector<int> min_input_shape,
+                                     std::vector<int> max_input_shape) {
+  PADDLE_ENFORCE_EQ(runtime_input_shape.size(), min_input_shape.size(),
+                    platform::errors::InvalidArgument(
+                        "TRT engine runtime input dims size(%d) inconsistent "
+                        "with the dynamic shape size(%d)",
+                        runtime_input_shape.size(), min_input_shape.size()));
+  auto is_input_shape_valid = [&](std::vector<int64_t> runtime_input_shape,
```
Comment (shangzhizhou) [inline paddle/fluid/operators/tensorrt/tensorrt_engine_op.h]: 同上

File: paddle/fluid/operators/tensorrt/tensorrt_engine_op.h [hunk 3]
```diff
@@ -76,6 +76,53 @@ static void RuntimeStaticShapeCheck(std::vector<int64_t> runtime_input_shape,
...
+  std::string runtime_input_shape_str = std::accumulate(
+      std::next(runtime_input_shape.begin()), runtime_input_shape.end(),
+      std::to_string(runtime_input_shape[0]), comma_fold);
+  std::string min_input_shape_str =
+      std::accumulate(std::next(min_input_shape.begin()), min_input_shape.end(),
+                      std::to_string(min_input_shape[0]), comma_fold);
+  std::string max_input_shape_str =
+      std::accumulate(std::next(max_input_shape.begin()), max_input_shape.end(),
+                      std::to_string(max_input_shape[0]), comma_fold);
+  PADDLE_ENFORCE_EQ(is_input_shape_valid(runtime_input_shape, min_input_shape,
```
Comment (shangzhizhou) [inline paddle/fluid/operators/tensorrt/tensorrt_engine_op.h:114]: 增加异常情况单测

File: paddle/fluid/operators/tensorrt/tensorrt_engine_op.h (hunk 1 above)
Comment (cryoco) [reply paddle/fluid/operators/tensorrt/tensorrt_engine_op.h]: done

File: paddle/fluid/operators/tensorrt/tensorrt_engine_op.h (hunk 3 above)
Comment (cryoco) [reply paddle/fluid/operators/tensorrt/tensorrt_engine_op.h:114]: done

File: python/paddle/fluid/tests/unittests/ir/inference/test_trt_dynamic_shape.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+        return {"data": np.random.random([1, 3, 64, 16]).astype("float32"), }
+
+    def test_check_output(self):
+        if core.is_compiled_with_cuda():
+            use_gpu = True
+            with self.assertRaises(Exception):
+                self.check_output_with_option(use_gpu)
+
+
+class TRTDynamicShapeOutOfBound2Test(TRTDynamicShapeTest):
```
Comment (shangzhizhou) [inline python/paddle/fluid/tests/unittests/ir/inference/test_trt_dynamic_shape.py:69]: 增加lower bound测试case

**Review by shangzhizhou — DISMISSED** [review]
LGTM

**Review by XieYunshen — APPROVED** [review]

**Pull request merged**